sdk/src/asset_handlers/font_io.rs

```diff
@@ -0,0 +1,644 @@
+        destination.write_u16::<BigEndian>(self.unitsPerEm)?;
+        // 0x14
+        destination.write_i64::<BigEndian>(self.created)?;
+        // 0x1c
+        destination.write_i64::<BigEndian>(self.modified)?;
+        // 0x24
+        destination.write_i16::<BigEndian>(self.xMin)?;
+        destination.write_i16::<BigEndian>(self.yMin)?;
+        // 0x28
+        destination.write_i16::<BigEndian>(self.xMax)?;
+        destination.write_i16::<BigEndian>(self.yMax)?;
+        // 0x2c
+        destination.write_u16::<BigEndian>(self.macStyle)?;
+        destination.write_u16::<BigEndian>(self.lowestRecPPEM)?;
+        // 0x30
+        destination.write_i16::<BigEndian>(self.fontDirectionHint)?;
+        destination.write_i16::<BigEndian>(self.indexToLocFormat)?;
+        // 0x34
+        destination.write_i16::<BigEndian>(self.glyphDataFormat)?;
+        // 0x36
+        destination.write_u16::<BigEndian>(0_u16)?;
+        // 0x38 - two bytes to get 54-byte 'head' up to nice round 56 bytes
+        Ok(())
+    }
+}
+
+/// Generic font table with unknown contents.
+#[derive(Debug)]
+pub struct TableUnspecified {
+    pub data: Vec<u8>,
+}
+
+/// Any font table.
+impl TableUnspecified {
+    /// Creates an unspecified table from the given stream.
+    pub fn make_from_reader<T: Read + Seek + ?Sized>(
+        reader: &mut T,
+        offset: u64,
+        size: usize,
+    ) -> core::result::Result<TableUnspecified, Error> {
+        let mut raw_table_data: Vec<u8> = vec![0; size];
+        reader.seek(SeekFrom::Start(offset))?;
+        reader.read_exact(&mut raw_table_data)?;
+        Ok(Self {
+            data: raw_table_data,
+        })
+    }
+
+    /// Compute the checksum
+    pub fn checksum(&self) -> Wrapping<u32> {
+        // TBD - this should never be called though - we only need to alter C2PA
+        // and head - should we panic!(), or just implement?
+        Wrapping(0x19283746)
+    }
+
+    /// Size of this table if it were serialized right now
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Serialize this table data to the given writer.
+    pub fn write<TDest: Write + ?Sized>(&self, destination: &mut TDest) -> Result<()> {
+        destination
+            .write_all(&self.data[..])
+            .map_err(|_e| Error::FontSaveError)?;
+        let limit = self.data.len() % 4;
+        if limit > 0 {
+            let pad: [u8; 3] = [0, 0, 0];
+            destination
+                .write_all(&pad[0..(4 - limit)])
+                .map_err(|_e| Error::FontSaveError)?;
+        }
+        Ok(())
+    }
+}
+
+/// Possible tables
+#[derive(Debug)]
+pub enum Table {
+    /// 'C2PA' table
+    C2PA(TableC2PA),
+    /// 'head' table
+    Head(TableHead),
+    /// any other table
+    Unspecified(TableUnspecified),
+}
+
+// TBD - This looks sort of like the CRTP from C++; do we want a Trait here
+// that *both* table *and* its value-types implement?
+impl Table {
```

Comment:
So.. rather than do this, I think what you'd want to do is:

```rust
trait Table {
  fn len(&self) -> usize;
  fn checksum(&self) -> Wrapping<u32>;
}
```

Then your types which are tables (`TableUnspecified`, `TableC2PA`, `TableHead`) would each implemented that trait  (`impl Table for TableUnspecified`).

This would allow you to get rid of the `pub enum Table` I think, allowing for your `SfntFont` (in `snft_io.rs`) to have:
```rust
struct Sfnt {
  tables: BTreeMap<SfntTag, Box<dyn Table>>
}
```

So you have a collection of "things which implement the `Table` trait".  This would then let you call `checksum()` and `len()` on anything in your table collection without caring what it is.

(note: you unfortunately can't just do `BTreeMap<SfntTag, Table>` because he wants to know how large each thing is in any collection, so you can't just use traits.  But if you do `Box<dyn Table>` you essentially get a vector of pointers to things which are that trait, which works)

Reply:
Split to [C2PA-305](https://monotype.atlassian.net/browse/C2PA-304).